certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,556 @@
+    bool executed_before = getSchedExeExecuted(index);
+
+    execute@withrevert(e, index);
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// Scheduled execution can only be executed by those supplied as executors during scheduling
+// https://prover.certora.com/output/40577/f84672d68a6d42498d6899d27b9288a1/?anonymousKey=cf4c6f18c3a8fad6fc659a983a8ed595775a061a
+rule cannotBecomeExecutorForAlreadyScheduledExecution(env e, method f) {
+    uint256 length = getSchedExeLength();
+    uint256 id;
+    address user;
+    require(length < max_uint256);
+    require(id < length);
+    bool isExecutor = isExecutor(id, user);
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    assert !isExecutor => !isExecutor(id, user),
+        "User became an executor after the scheduled execution has been scheduled.";
+}
+
+
+// STATUS - verified
+// Only execution, that has not yet been cancelled or executed can be executed
+// and only if it's executableAt is not greater than current timestamp.
+// Also, when execution happens, the caller must be executor.
+// Also, executor can execute execution, which is executable (`canExecute`)
+// https://prover.certora.com/output/40577/0b63385b7d2c4389acfcd10e8e4babfa/?anonymousKey=602c9b0b12b1e83f7625b142d5244a03f6ef6a3c
+rule whoCanExecuteAndWhatIsExecutable(env e) {
+    uint256 length = getSchedExeLength();
+    uint256 id;
+    require(id < length);
+
+    bool canExecute = canExecute(e, id);
+    bool isExecutor = isExecutor(id, e.msg.sender);
+    bool isProtected = getSchedExeProtected(id);
+
+    assert (
+        !getSchedExeExecuted(id) &&
+        !getSchedExeCancelled(id) &&
+        getSchedExeExecutableAt(id) < e.block.timestamp
+    ) => canExecute;
+
+    bool executedBefore = getSchedExeExecuted(id);
+
+    execute(e, id);
+
+    bool executedAfter = getSchedExeExecuted(id);
+
+    assert !executedBefore && executedAfter => isExecutor || !isProtected;
+    assert canExecute && isExecutor => executedAfter;
+
+    cancel@withrevert(e, id);
+    bool reverted = lastReverted;
+    assert executedAfter => reverted;
+}
+
+// STATUS - verified
+// https://prover.certora.com/output/40577/932a5378e8b043c2ac0a07b82dd2e3f3/?anonymousKey=9b663bca7afe0aff0559fe37a922e54193765410
+rule cancelerCanCancel(env e) {
+    uint256 length = getSchedExeLength();
+    uint256 id;
+    require(id < length);
+    bool isCanceler = isCanceler(id, e.msg.sender);
+
+    bool canceledBefore = getSchedExeCancelled(id);
+    bool executed = getSchedExeExecuted(id);
+
+    cancel(e, id);
+
+    bool canceledAfter = getSchedExeCancelled(id);
+
+    assert canceledAfter;
+    assert !canceledBefore => isCanceler;
+
+    execute@withrevert(e, id);
+    bool reverted = lastReverted;
+    assert canceledAfter => reverted;
+}
+
+
+// STATUS - verified
+// `isExecutor` can only be changed by `_executionHelper` and
+// it can only happen by executing a non-schedule function.
```

Comment:
I think it's incorrect. You can add/create an executor when you schedule execution. Anybody with scheduling permission can schedule.

Reply:
Yes, this description was nonsense and does not correspond to the asserts in the rule. I fixed it.